File: src/main/test/com/edgelab/hospital/IntegrationTests.java
package com.edgelab.hospital;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class IntegrationTests {

    private static PrintStream OLD_OUT;

    private static ByteArrayOutputStream BAOS;

    @BeforeEach
    void redirectPrintStream() {
        BAOS = new ByteArrayOutputStream();
        PrintStream redirectOut = new PrintStream(BAOS);
        OLD_OUT = System.out;
        System.setOut(redirectOut);
    }

    @AfterEach
    void setPrintStreamBack() {
        System.out.flush();
        System.setOut(OLD_OUT);
    }

    @Test
    void noPatientsNoDrugs() {
        Application.main(new String[]{"", ""});
        assertEquals("F:0,H:0,D:0,T:0,X:0", BAOS.toString().replaceAll("([\\r\\n])", ""));
    }

    @Test
    void noPatientsMultiDrugs() {
        Application.main(new String[]{"", "P,As"});
        assertEquals("F:0,H:0,D:0,T:0,X:0", BAOS.toString().replaceAll("([\\r\\n])", ""));
    }

    @Test
    void onePatientNoDrug() {
        Application.main(new String[]{"F", ""});
        assertEquals("F:1,H:0,D:0,T:0,X:0", BAOS.toString().replaceAll("([\\r\\n])", ""));
    }

    @Test
    void onePatientOneDrug() {
        Application.main(new String[]{"F", "P"});
        assertEquals("F:0,H:1,D:0,T:0,X:0", BAOS.toString().replaceAll("([\\r\\n])", ""));
    }

    @Test
    void multiPatientsNoDrugs() {
        Application.main(new String[]{"D,D", ""});
        assertEquals("F:0,H:0,D:0,T:0,X:2", BAOS.toString().replaceAll("([\\r\\n])", ""));
    }

    @Test
    void multiPatientsMultiDrugs() {
        Application.main(new String[]{"F,D,T,T", "P,An"});
        assertEquals("F:0,H:3,D:0,T:0,X:1", BAOS.toString().replaceAll("([\\r\\n])", ""));
    }

}

File: src/test/kotlin/com/edgelab/hospital/IntegrationTests.kt
package com.edgelab.hospital

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.ByteArrayOutputStream
import java.io.PrintStream

class IntegrationTests {

  companion object {
    private var BAOS: ByteArrayOutputStream? = null
  }

  @BeforeEach
  fun redirectPrintOutStream() {
    BAOS = ByteArrayOutputStream()
    System.setOut(PrintStream(BAOS!!))
  }

  @Test
  fun invalidArguments() {
    assertThatExceptionOfType(IllegalArgumentException::class.java)
      .isThrownBy{ main(arrayOf("")) }
  }

  @Test
  fun emptyArguments() {
    main(arrayOf("", ""))
    assertThat(BAOS.toString().replace("([\\r\\n])".toRegex(), "")).isEqualTo("F:0,H:0,D:0,T:0,X:0")
  }

  @Test
  fun noPatientsMultiDrugs() {
    main(arrayOf("", "P,As"))
    assertThat(BAOS.toString().replace("([\\r\\n])".toRegex(), "")).isEqualTo("F:0,H:0,D:0,T:0,X:0")
  }

  @Test
  fun onePatient() {
    main(arrayOf("D", ""))
    assertThat(BAOS.toString().replace("([\\r\\n])".toRegex(), "")).isEqualTo("F:0,H:0,D:0,T:0,X:1")
  }

  @Test
  fun onePatientNoDrug() {
    main(arrayOf("F", ""))
    assertThat(BAOS.toString().replace("([\\r\\n])".toRegex(), "")).isEqualTo("F:1,H:0,D:0,T:0,X:0")
  }

  @Test
  fun onePatientOneDrug() {
    main(arrayOf("F", "P"))
    assertThat(BAOS.toString().replace("([\\r\\n])".toRegex(), "")).isEqualTo("F:0,H:1,D:0,T:0,X:0")
  }

  @Test
  fun multiPatientsNoDrugs() {
    main(arrayOf("D,D", ""))
    assertThat(BAOS.toString().replace("([\\r\\n])".toRegex(), "")).isEqualTo("F:0,H:0,D:0,T:0,X:2")
  }

  @Test
  fun multiPatientsMultiDrugs() {
    main(arrayOf("F,D,T,T", "P,An"))
    assertThat(BAOS.toString().replace("([\\r\\n])".toRegex(), "")).isEqualTo("F:0,H:3,D:0,T:0,X:1")
  }

}
